src/link_manager/LinkManagerStateMachineActiveActive.cpp

```diff
@@ -0,0 +1,934 @@
+{
+    if (mLabel != label) {
+        mLabel = label;
+        mMuxPortPtr->setMuxLinkmgrState(label);
+
+        MUXLOGWARNING(
+            boost::format("%s: Linkmgrd state is: %s %s") %
+            mMuxPortConfig.getPortName() %
+            mMuxStateName[ms(mCompositeState)] %
+            mLinkHealthName[static_cast<int>(label)]
+        );
+    }
+}
+
+//
+// ---> enterLinkProberState(CompositeState &nextState, link_prober::LinkProberState::Label label);
+//
+// force LinkProberState to switch state
+//
+void ActiveActiveStateMachine::enterLinkProberState(
+    CompositeState &nextState,
+    link_prober::LinkProberState::Label label
+)
+{
+    mLinkProberStateMachinePtr->enterState(label);
+    ps(nextState) = label;
+}
+
+//
+// ---> enterMuxState(CompositeState &nextState, mux_state::MuxState::Label label);
+//
+// force MuxState to switch state
+//
+void ActiveActiveStateMachine::enterMuxState(
+    CompositeState &nextState,
+    mux_state::MuxState::Label label
+)
+{
+    mMuxStateMachine.enterState(label);
+    ms(nextState) = label;
+}
+
+//
+// ---> enterLinkState(CompositeState &nextState, link_state::LinkState::Label label);
+//
+// force LinkState to switch state
+//
+void ActiveActiveStateMachine::enterLinkState(
+    CompositeState &nextState,
+    link_state::LinkState::Label label
+)
+{
+    mLinkStateMachine.enterState(label);
+    ls(nextState) = label;
+}
+
+//
+// ---> enterPeerMuxState(mux_state::MuxState::Label label);
+//
+// force peer MuxState to switch state
+//
+void ActiveActiveStateMachine::enterPeerMuxState(mux_state::MuxState::Label label)
+{
+    mPeerMuxState = label;
+}
+
+//
+// ---> enterPeerLinkProberState(link_prober::LinkProberState::Label label);
+//
+// force peer LinkProberState to switch state
+//
+void ActiveActiveStateMachine::enterPeerLinkProberState(link_prober::LinkProberState::Label label)
+{
+    mLinkProberStateMachinePtr->enterPeerState(label);
+    mPeerLinkProberState = label;
+}
+
+//
+// ---> switchMuxState(CompositeState &nextState, mux_state::MuxState::Label label, bool forceSwitch);
+//
+// switch MUX to the target state
+//
+void ActiveActiveStateMachine::switchMuxState(
+    CompositeState &nextState,
+    mux_state::MuxState::Label label
+)
+{
+    if (mMuxPortConfig.getMode() == common::MuxPortConfig::Mode::Auto) {
+        MUXLOGWARNING(
+            boost::format("%s: Switching MUX state to '%s'") %
+            mMuxPortConfig.getPortName() %
+            mMuxStateName[label]
+        );
+        if (label == mux_state::MuxState::Label::Standby) {
+            // suspend heartbeat to help peer ToR to toggle
+            mSuspendTxFnPtr(mMuxPortConfig.getLinkWaitTimeout_msec());
+        }
+        enterMuxState(nextState, label);
+        mMuxStateMachine.setWaitStateCause(mux_state::WaitState::WaitStateCause::SwssUpdate);
+        mMuxPortPtr->setMuxState(label);
+        mDeadlineTimer.cancel();
+        startMuxWaitTimer();
+    } else {
+        probeMuxState();
```

Comment:
if in non-auto mode, and probing back a mux state that is different than mux mode, will we switch to match mux mode?

Reply:
I guess the question is, is mux mode a "must-be" or a "preference"?

Reply:
Currently, the mux mode is more like a one-time-off setting, after setting the mux mode, `linkmgrd` now assumes the toggle must be a success, and won't react to any more state changes.

Reply:
Will it make more sense in `active-active` to keep mux state aligned with mux mode? But we can leave it as same as `active-standby` (one-time setting) for now.